hera_mc/correlator.py

```diff
@@ -0,0 +1,249 @@
+        """
+        Create a new correlator control command object.
+
+        Parameters:
+        ------------
+        time: astropy time object
+            astropy time object for time command was sent.
+        command: string
+            one of the keys in command_dict (e.g. 'take_data',
+            'phase_switching_on', 'phase_switching_off', 'restart')
+        """
+        if not isinstance(time, Time):
+            raise ValueError('time must be an astropy Time object')
+        corr_time = floor(time.gps)
+
+        if command not in list(command_dict.keys()):
+            raise ValueError('command must be one of: ' + ', '.join(list(command_dict.keys()))
+                             + '. command is actually {}'.format(command))
+
+        return cls(time=corr_time, command=command)
+
+
+def _get_integration_time(acclen_spectra, correlator_redis_address=DefaultRedisAddress):
+    """
+    gets the integration time in seconds for a given acclen in spectra
+
+    """
+    import hera_corr_cm
+
+    corr_cm = hera_corr_cm.HeraCorrCM(redishost=correlator_redis_address)
+
+    # TODO: call actual function that Jack will make to convert acclen_spectra to
+    # integration time in seconds (this is dummy code for now)
+    return corr_cm.get_integration_time(acclen_spectra)
+
+
+class CorrelatorTakeDataArguments(MCDeclarativeBase):
+    """
+    Definition of correlator take_data arguments table.
+
+    Records the arguments passed to the correlator `take_data` command.
+
+    time: gps time of the take_data command, floored (BigInteger, part of primary_key).
+    starttime_sec: gps time to start taking data, floored (BigInteger)
+    starttime_ms: milliseconds to add to starttime_sec to set correlator start time.
+    duration: Duration to take data for in seconds. After this time, the
+        correlator will stop recording.
+    acclen_spectra: Accumulation length in spectra.
+    integration_time: Accumulation length in seconds, converted from acclen_spectra.
+        The conversion is non-trivial and depends on the correlator settings.
+    tag: Tag which will end up in data files as a header entry
+    """
+    __tablename__ = 'correlator_take_data_arguments'
+    time = Column(BigInteger, primary_key=True)
+    command = Column(String, primary_key=True)
+    starttime_sec = Column(BigInteger)
+    starttime_ms = Column(Integer)
+    duration = Column(Float)
+    acclen_spectra = Column(Integer)
+    integration_time = Column(Float)
+    tag = Column(String)
+
+    # TODO: this Foreign key arrangement requires an unneccesary column on this
+    # object (command) because this table only has rows when command='take_data'
+    # Is there a way to specify the constraint without this dummy column?
+    __table_args__ = (ForeignKeyConstraint(['time', 'command'],
+                                           ['correlator_control_command.time',
+                                            'correlator_control_command.command']), {})
```

Comment:
@pkgw do you know of a way to set up this foreign key constraint without this unnecessary command column?

Reply:
No, I don't. This is not an area where I have much expertise at all, but I doubt it is possible to do so. My guess is that if you don't want the `command` column, you have to drop the explicit statement of the foreign key constraint. Which might be fine? It's kind of a judgment call, but given that these tables are likely to be append-only, I think it would be fairly safe to *not* formally enforce the constraint.